### PR TITLE
Add loop contracts and harness for `slice::is_ascii`

### PR DESCRIPTION
**Pull request by qinheping**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/lib.rs [hunk 1]
```diff
@@ -230,6 +230,7 @@
 #![feature(unboxed_closures)]
 #![feature(unsized_fn_params)]
 #![feature(with_negative_coherence)]
+#![feature(proc_macro_hygiene)]
```
Comment (celinval) [inline library/core/src/lib.rs:233]:
```suggestion
#![cfg_attr(kani, feature(proc_macro_hygiene))]
```

File: library/core/src/slice/ascii.rs [hunk 2]
```diff
@@ -432,3 +439,28 @@ const fn is_ascii(s: &[u8]) -> bool {
...
+#[cfg(kani)]
+#[unstable(feature = "kani", issue = "none")]
+pub mod verify {
+    use super::*;
+
+    #[kani::proof]
+    #[kani::unwind(8)]
+    pub fn check_is_ascii() {
+        if kani::any() {
+            // TODO: ARR_SIZE can be `std::usize::MAX` with cbmc argument
```
Comment (celinval) [inline library/core/src/slice/ascii.rs:452]: I believe Rust has a constraint that is has to be up to `isize::MAX`.

**Review by celinval — APPROVED** [review]

**Review by carolynzech — APPROVED** [review]

File: library/core/src/lib.rs (hunk 1 above)
Comment (celinval) [reply library/core/src/lib.rs:233]: please add a comment instead.

File: library/core/src/slice/ascii.rs [hunk 3]
```diff
@@ -398,6 +401,10 @@ const fn is_ascii(s: &[u8]) -> bool {
     // Read subsequent words until the last aligned word, excluding the last
     // aligned word by itself to be done in tail check later, to ensure that
     // tail is always one `usize` at most to extra branch `byte_pos == len`.
+    #[safety::loop_invariant(byte_pos <= len 
+                            && byte_pos >= offset_to_aligned
+                            && word_ptr as usize >= start as usize + offset_to_aligned
+                            && (byte_pos - offset_to_aligned) == (word_ptr as usize - start as usize - offset_to_aligned))]
```
[the remark this answers is not in the document]
Comment (celinval) [reply library/core/src/slice/ascii.rs:407]: also, please replace all the `as usize` with `addr()`.

File: library/core/src/slice/ascii.rs [hunk 4]
```diff
@@ -432,3 +439,28 @@ const fn is_ascii(s: &[u8]) -> bool {
...
     !contains_nonascii(last_word)
 }
+
+#[cfg(kani)]
+#[unstable(feature = "kani", issue = "none")]
+pub mod verify {
+    use super::*;
+
+    #[kani::proof]
+    #[kani::unwind(8)]
```
Comment (zhassan-aws) [inline library/core/src/slice/ascii.rs:449]: Same question on the unwind here.

**Review by celinval — COMMENTED** [review]
trigger CR check

File: library/core/src/slice/ascii.rs (hunk 4 above)
Comment (qinheping) [reply library/core/src/slice/ascii.rs:449]: 'is_ascii' calls 'is_ascii_simple' to handle base case where the loop in 'is_ascii_simple' execute at most sizeof(usize) number of times. So we unwind it 8 times.

**Review by celinval — COMMENTED** [review]
trigger approval checks

**Pull request merged**